### PR TITLE
Implement translation of ## operator.

### PR DESCRIPTION
**Pull request by ciechowoj**
I know that I was told to avoid PR that mixes different things, but maintaining multiple branches is painful. For my defense, the features have quite small implementations and are in self-contained commits.
Features:
+ Implement translation of ## operator.
+ Fix translation of function pointer typedefs.
+ Fix formatting bug (details in unittest).
+ Fix translation of array typedef.

### CODE REVIEW

File: unit_tests/MacroTranslTests.d [hunk 1]
```diff
@@ -599,3 +599,41 @@ enum FOO = Bar.sizeof;
...
+unittest
+{
+    assertTranslates(q"C
+#define FOO_CONCAT(prefix, name) prefix ## name
+C", q"D
+extern (C):
+
+extern (D) string FOO_CONCAT(T0, T1)(auto ref T0 prefix, auto ref T1 name)
+{
+    return prefix ~ name;
```
Comment (jacob-carlborg) [inline unit_tests/MacroTranslTests.d]: It's possible to pass in integers to a macro like declared above. `~` will not work for integers.

Comment (ciechowoj) [inline unit_tests/MacroTranslTests.d]: Good point. I've implemented improved version.

**Comment by ciechowoj** [comment]
Is there anything I can improve here?

**Comment by jacob-carlborg** [comment]
I just think I have not had the time to to a proper review yet.

**Comment by jacob-carlborg** [comment]
Is this PR up to date?

File: dstep/translator/MacroDefinition.d [hunk 2]
```diff
@@ -436,24 +441,68 @@ class StringConcat : Expression
...
+        this.subexprs = subexprs;
+    }
+
+    override string translate(Context context, Set!string params, ref Set!string imports)
+    {
+        import std.algorithm.iteration : map;
+        import std.array : join;
+        import std.format : format;
+
+        string fun(Expression subexpr)
```
Comment (ciechowoj) [inline dstep/translator/MacroDefinition.d]: A shortcut for `function`.

**Comment by ciechowoj** [comment]
Yes.

File: dstep/translator/MacroDefinition.d (hunk 2 above)
Comment (jacob-carlborg) [inline dstep/translator/MacroDefinition.d]: Ideally you should try to come up with a name that actually describes that the function does, just as if it was a non-nested function.

Comment (ciechowoj) [inline dstep/translator/MacroDefinition.d]: Fixed.

**Comment by jacob-carlborg** [comment]
👌

**Comment by jacob-carlborg** [comment]
Can you please update the changelog.

**Comment by ciechowoj** [comment]
Updated.

**Comment by jacob-carlborg** [comment]
Thanks.

**Comment by jacob-carlborg** [comment]
No more open pull reqeusts by you 😃.

**Comment by jacob-carlborg** [comment]
BTW, nice work 👍.

**Comment by ciechowoj** [comment]
> BTW, nice work :+1:.

Thanks : )

**Pull request merged**